qiskit/transpiler/runningpassmanager.py

```diff
@@ -102,20 +102,27 @@ def _normalize_flow_controller(self, flow_controller):
                 raise TranspilerError("The flow controller parameter %s is not callable" % name)
         return flow_controller
 
-    def run(self, circuit, output_name=None, callback=None):
+    def run(self, circuit, output_name=None, callback=None, return_dag=False):
         """Run all the passes on a QuantumCircuit
 
         Args:
-            circuit (QuantumCircuit): circuit to transform via all the registered passes
+            circuit (QuantumCircuit | DAGCircuit): circuit to transform via all the
+                                                   registered passes
             output_name (str): The output circuit name. If not given, the same as the
                                input circuit
-            callback (callable): A callback function that will be called after each pass execution.
+            callback (callable): A callback function that will be called after each
+                                 pass execution.
+            return_dag (bool): If True return DAGCircuit as output instead of converting
+                               back to a QuantumCircuit.
+
         Returns:
             QuantumCircuit: Transformed circuit.
         """
-        name = circuit.name
-        dag = circuit_to_dag(circuit)
-        del circuit
+        if isinstance(circuit, DAGCircuit):
+            dag = circuit
+        else:
+            dag = circuit_to_dag(circuit)
+            del circuit
```

Comment:
The aim of the `del` is to reduce the references to the input circuit so that when it's overwritten, we're not keeping the original one alive here.  If the `if` branch is taken, then the name `circuit` persists, so there are least two living references to the input within this function, so even when `dag` is reassigned, there's still a reference.  To maintain the spirit of the original, the `del` should be unconditional.

_That said_, I don't think the `del` is actually doing anything much, because the outer caller of `RunningPassManager.run` still holds a reference to the circuit which isn't released, so the circuit can't get garbage collected until this function returns at the very earliest anyway.